### PR TITLE
Fix some threading docs

### PR DESCRIPTION
**Pull request by kouvel**
- Ported https://github.com/dotnet/dotnet-api-docs/pull/4887 from 5.0-rc2 to master
- Added exception case for `ExecutionContext.Restore`
- Improved docs for named wait handles
- Made a small update to samples for interlocked decrement/increment
- Updated some `ExecutionContext` docs for .NET Core

Fixes https://github.com/dotnet/runtime/issues/35519
Fixes https://github.com/dotnet/runtime/issues/35518
Fixes https://github.com/dotnet/dotnet-api-docs/issues/4585
Fixes https://github.com/dotnet/dotnet-api-docs/issues/4448

### CODE REVIEW

**Comment by kouvel** [comment]
Made a couple of fixes to note some subtle differences

**Comment by kouvel** [comment]
Rebased to fix conflicts. The last two commits are new.

**Comment by kouvel** [comment]
Not sure what the validation warning is referring to, I don't see any non-unique constructor signatures

**Comment by kouvel** [comment]
I think this is ready to merge. It looks like the approval has disappeared, maybe after I closed/reopened the issue to work around a CI validation issue above. @mangod9 could you please approve again?

**Review by mangod9 — APPROVED** [review]

**Comment by mangod9** [comment]
Approved again, but perhaps it requires approval from the docs team.

**Comment by kouvel** [comment]
Thanks! @gewarren could you please also take a quick look?

**Review by gewarren — COMMENTED** [review]
I left a few suggestions, that if accepted, should be updated across all the files.
Also, it looks like the build warning about the duplicate .ctor is bogus - all 5 constructors look unique to me. I'll inquire with the team responsible for that.

File: xml/System.Threading/EventWaitHandleAcl.xml [hunk 1]
```diff
@@ -53,20 +53,29 @@
           <format type="text/markdown"><![CDATA[
 
 ## Remarks
+ The `name` may be prefixed with `Global\` or `Local\` to specify a namespace. When the `Global` namespace is specified, the synchronization object may be shared with any processes on the system. When the `Local` namespace is specified, which is also the default when no namespace is specified, the synchronization object may be shared with processes in the same session. On Windows, a session is a login session, and services typically run in a different non-interactive session. On Unixes, each shell has its own session. Session-local synchronization objects may be appropriate for synchronizing between processes with a parent/child relationship where they all run in the same session. For more information about synchornization object names on Windows, see https://docs.microsoft.com/windows/win32/sync/object-names.
```
[the remark this answers is not in the document]
Comment (kouvel) [reply xml/System.Threading/EventWaitHandleAcl.xml]: It looks like "Unixes" is the common plural form and can also be used to refer to Unix-like systems. I think your suggestion is more clear, changed to "On Unix-like operating systems...".

File: xml/System.Threading/EventWaitHandleAcl.xml [hunk 2]
```diff
@@ -53,20 +53,29 @@
...
 
 -or-
 
-.NET Framework only: The <paramref name="mode" /> enum value was out of legal range.</exception>
-        <exception cref="T:System.ArgumentOutOfRangeException">The <paramref name="mode" /> enum value was out of legal range.</exception>
-        <exception cref="T:System.IO.DirectoryNotFoundException">Could not find a part of the path specified in <paramref name="name" />.</exception>
-        <exception cref="T:System.Threading.WaitHandleCannotBeOpenedException">A system-wide synchronization event with the provided <paramref name="name" /> was not found.
- -or- An <see cref="T:System.Threading.EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="T:System.Threading.EventWaitHandle" /> of a different type might have the same name.</exception>
+.NET Framework only: <paramref name="name" /> is longer than MAX_PATH (260 characters).</exception>
+        <exception cref="T:System.ArgumentOutOfRangeException">The <paramref name="mode" /> enum value was out of legal range. In some cases <see cref="T:System.ArgumentException"/> is thrown, so use that instead.</exception>
```
[the remark this answers is not in the document]
Comment (kouvel) [reply xml/System.Threading/EventWaitHandleAcl.xml]: I didn't want to suggest catching this exception, I just removed that part and just noted the different exception that may be thrown. Also removed another similar usage for `DirectoryNotFoundException`.

File: xml/System.Threading/EventWaitHandleAcl.xml [hunk 3]
```diff
@@ -53,20 +53,29 @@
...
+
+-or-
+
+There was some other error. The `HResult` property may provide more information.</exception>
+        <exception cref="T:System.IO.DirectoryNotFoundException">Windows only: <paramref name="name" /> specified an unknown namespace. See https://docs.microsoft.com/windows/win32/sync/object-names for more information. For cross-platform compatibility, use <see cref="T:System.IO.IOException"/> instead.</exception>
+        <exception cref="T:System.IO.PathTooLongException">The <paramref name="name" /> is too long. Length restrictions may depend on the operating system or configuration.</exception>
+        <exception cref="T:System.Threading.WaitHandleCannotBeOpenedException">A synchronization object with the provided <paramref name="name" /> cannot be created. A synchronization object of a different type might have the same name.</exception>
+        <exception cref="T:System.UnauthorizedAccessException">The named event exists, but the user does not have the desired security access.</exception>
```
[the remark this answers is not in the document]
Comment (kouvel) [reply xml/System.Threading/EventWaitHandleAcl.xml:78]: It's referring to the access requested in the parameter. I probably would have written "requested security access", just chose the same wording used in a few other places.

**Comment by kouvel** [comment]
@gewarren could you please take a look at the changes above?

**Review by gewarren — APPROVED** [review]

**Pull request merged**